README.md
# grphp-statsd - StatsD interceptor for grphp

StatsD interceptor for measuring [grphp](https://github.com/bigcommerce/grphp) client requests.

## Installation

```bash
composer require bigcommerce/grphp-statsd
```

## Configuration

```php
$config = [
    'host' => 'my.statsd.service',
    'port' => 8125,
    
    // optional config options
    'namespace' => 'myapp',
    'sample_rate' => 1.0, // 0->1.0 scale. 1.0 is all requests.
    'mtu' => 1500, // MTU rate
    'persistent' => false, // keep connection persistent
    'timeout' => 600, // in seconds
    'tcp' => false, // true uses TCP instead of UDP
];
$client->addInterceptor(new \Grphp\StatsD\Interceptor($config));
```

## License

Copyright (c) 2017-present, BigCommerce Pty. Ltd. All rights reserved 

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated 
documentation files (the "Software"), to deal in the Software without restriction, including without limitation the 
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit 
persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the 
Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE 
WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR 
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR 
OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
#Changelog for grphp-statsd.

###0.2.0

* Add `PHP` 8 support
* Add `phpunit` 9 support
* Update `grphp` dependency to 3.0
* Move from `travis` to `circleci`

###0.1.1

* Update grphp dependency to 0.3.8

###0.1.0

* Better namespaced key support
* Added full unit tests
* Added code coverage

###0.0.3

* Fix composer dependencies
* Fix timer to properly measure in ms

###0.0.2

* grphp 0.3.0 support

###0.0.1

* Initial commit


tests/bootstrap.php
<?php
/**
 * Copyright (c) 2017-present, BigCommerce Pty. Ltd. All rights reserved
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
error_reporting(E_ALL | E_STRICT);
require dirname(__DIR__).'/vendor/autoload.php';
date_default_timezone_set('UTC');
require dirname(__DIR__) . '/tests/Support/BaseTest.php';


tests/Support/BaseTest.php
<?php
/**
 * Copyright (c) 2017-present, BigCommerce Pty. Ltd. All rights reserved
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
namespace Grphp\StatsD\Test;

use PHPUnit\Framework\TestCase;

class BaseTest extends TestCase
{
}


tests/Support/TestStub.php
<?php
/**
 * Copyright (c) 2017-present, BigCommerce Pty. Ltd. All rights reserved
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
namespace Grphp\Statsd\Test;

class BaseStub extends \Grpc\BaseStub
{
    public function __construct()
    {
    }
}

class TestClient extends BaseStub
{
}

class CallStatus
{
    public $code = 0;
    public $details = '';
    public $metadata = [];

    public function __construct($code = 0, $details = '', $metadata = [])
    {
        $this->code = $code;
        $this->details = $details;
        $this->metadata = $metadata;
    }
}


src/Grphp/StatsD/Interceptor.php
<?php
/**
 * Copyright (c) 2017-present, BigCommerce Pty. Ltd. All rights reserved
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
namespace Grphp\StatsD;

use Grphp\Client\Interceptors\Base;

class Interceptor extends Base
{
    public function __construct(array $options = [])
    {
        parent::__construct($options);
    }

    /** @var \Domnikl\Statsd\Client $client */
    private $client;

    /**
     * @param callable $callback
     * @return \Grphp\Client\Response
     */
    public function call(callable $callback)
    {
        $start = microtime(true);
        /** @var \Grphp\Client\Response $response */
        $response = $callback();
        $elapsed = microtime(true) - $start;
        $elapsed = round($elapsed * 1000.00, 4);

        $this->measure($elapsed, $response->isSuccess());
        return $response;
    }

    /**
     * Measure the response in statsd
     *
     * @param float $elapsed time elapsed in ms
     * @param bool $success if the response is a successful one or not
     */
    private function measure($elapsed, $success = true)
    {
        $k = $this->key();
        $cl = $this->client();
        $cl->timing($k, $elapsed);
        $cl->increment($k);
        if ($this->option('gauge', false)) {
            $cl->gauge($k, $elapsed);
        }
        if ($success) {
            $cl->increment($k . '.success');
        } else {
            $cl->increment($k . '.failure');
        }
    }

    /**
     * @return string
     */
    private function key()
    {
        return $this->serviceKey() . '.' . $this->methodKey();
    }

    /**
     * @return mixed
     */
    private function methodKey()
    {
        return str_replace('\\', '.', strtolower($this->getMethod()));
    }

    /**
     * @return string
     */
    private function serviceKey()
    {
        $s = get_class($this->getStub());
        $s = str_replace('\\', '.', $s);
        $s = explode('.', $s);
        foreach ($s as &$c) {
            $c = strtolower(preg_replace('/(?<!^)[A-Z]/', '_$0', $c));
        }
        return implode('.', $s);
    }

    /**
     * @return \Domnikl\Statsd\Client
     */
    private function client()
    {
        if (empty($this->client)) {
            $this->client = $this->option('client');
            if (empty($this->client)) {
                $this->client = ClientFactory::build($this->getOptions());
            }
        }
        return $this->client;
    }

    /**
     * @param string $key
     * @param mixed $default
     * @return mixed
     */
    private function option($key, $default = null)
    {
        $config = $this->getOptions();
        return array_key_exists($key, $config) ? $config[$key] : $default;
    }
}


src/Grphp/StatsD/ClientFactory.php
<?php
/**
 * Copyright (c) 2017-present, BigCommerce Pty. Ltd. All rights reserved
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
 * documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to
 * permit persons to whom the Software is furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
 * Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
 * WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
 * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
 * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
namespace Grphp\StatsD;

use Domnikl\Statsd\Client as StatsClient;
use Domnikl\Statsd\Connection\TcpSocket;
use Domnikl\Statsd\Connection\UdpSocket;

/**
 * Generates StatsD clients
 *
 * @package Grphp\StatsD
 */
class ClientFactory
{
    public function __construct(array $config = [])
    {
        $this->config = $config;
    }

    /**
     * Build a new statsd client
     *
     * @param array $config
     * @return StatsClient
     */
    public static function build(array $config = [])
    {
        $factory = new static($config);
        return $factory->generate();
    }

    /**
     * @return StatsClient
     */
    public function generate()
    {
        $tcp = $this->option('tcp', false);
        $port = $this->option('port', 8125);
        $hostname = $this->option('host', '127.0.0.1');
        $namespace = $this->option('namespace', '');
        $timeout = $this->option('timeout');
        $persistent = $this->option('persistent', false);
        $mtu = $this->option('mtu', 1500);
        $sampleRate = $this->option('sample_rate', 1.0);

        if ($tcp) {
            $connection = new TcpSocket($hostname, $port, $timeout, $persistent, $mtu);
        } else {
            $connection = new UdpSocket($hostname, $port, $timeout, $persistent, $mtu);
        }
        return new StatsClient($connection, $namespace, $sampleRate);
    }

    /**
     * Get a set config option with a default
     *
     * @param string $k
     * @param mixed $default
     * @return mixed
     */
    private function option($k, $default = null)
    {
        return array_key_exists($k, $this->config) ? $this->config[$k] : $default;
    }
}
